### DISCUSSION TITLE
Can a grammar be defined such that only a particular integer is allowed?

### DISCUSSION THREAD

**Discussion by Beakerboy** [antlr/grammars-v4#4128] [Q&A]
In VBA, the following are valid:

```vba
GoTo 1
GoTo 2
On Error GoTo 1
On Error GoTo -1
```

But these are both invalid:
```vba
GoTo -1
On Error GoTo -2
```

In the official VBA syntax, an `INTEGER` is only a sequence of digits, and this is what my current grammar is expecting in these cases. In most mathematical cases, the `MINUS` is a unary operator defined in the parser.

I’m struggling with trying to figure out how to either define a `MINUS_ONE` token, or a Lexer rule that will only match “-1”. Should I instead just add `MINUS INTEGER” to the parser rule and let a Lint tool sort it out later if the minus exists and the value of the integer is anything other than a “1”?

**Comment by kaby76** [comment]
This is an example of when you need to move the syntax into a static semantics analysis. You should not be adding these checks into the grammar. Separate context-free syntax and semantic checks into completely different docs: the CFG is in the Antlr4 .g4 file; semantic checks need to be defined elsewhere. You can have a constraint system called post-parse on the entire tree, or during the parse on partial trees.

For example, with the [grammars-v4/vba](https://github.com/antlr/grammars-v4/tree/d32a2973696e956be86cd581d6e36293701ded4e/vba) grammar, to handle `On Error GoTo -2`, I would specify the constraint using XPath.
```
06/07-08:45:50 ~/issues/g4-current/vba/Generated-CSharp
$ cat xxx
Public Sub Module()
    ON ERROR GOTO -2
End Sub
06/07-08:45:52 ~/issues/g4-current/vba/Generated-CSharp
$ trparse -l xxx | trxgrep ' //onErrorStmt[ON_ERROR]/valueStmt/literal/SHORTLITERAL[text() = "-2"]' | trcaret
CSharp 0 xxx success 0.1587228
L2:     ON ERROR GOTO -2
                      ^
06/07-08:45:57 ~/issues/g4-current/vba/Generated-CSharp
$
```
Note, the vba grammar defines a [SHORTLITERAL](https://github.com/antlr/grammars-v4/blob/d32a2973696e956be86cd581d6e36293701ded4e/vba/vba.g4#L1879-L1881) with `'-'`. I would not define the integer literal with an operator.

**Reply by Beakerboy** [reply]
Thanks for this. I am working on a new grammar for VBA 7.1 based on the official Microsoft language specification. The new [INTEGERLITERAL](https://github.com/Beakerboy/grammars-v4/blob/coverage/vba/vba7_1/vbaLexer.g4#L1081) definition is a little different than what was put together for the existing VBA6 grammar.

I’m using this new grammar to build a Linting and static analysis tool, so I’ll make sure to place those checks there.
